### ISSUE TITLE
Process with Praat

### ISSUE THREAD

**Issue by robertfromont** [nzilbb/labbcat-R#3]
A function to provide acoustic measures from Praat for tokens in a results.csv file (or a variable containing tokens)

**Comment by robertfromont** [comment]
One tricky thing is that the *process with praat* page in LaBB-CAT has a bunch of options that would need to be specifiable somehow in the R function, including:

  * the start and end time columns
  * the window offset (surrounding context to extract)
  * which measurements to make, including their options:
    + formants
        - F1, F2, and/or F3
        - how many sample points and where they should be in the interval
        - max formants for 'female' and 'male' participants
        - the exact Praat command to use
    + pitch
        - min, mean, and/or max
        - pitch floor for 'female' and 'male' participants
        - pitch ceiling for 'female' and 'male' participants
        - voicing threshold for 'female' and 'male' participants
        - the exact praat command to use
    + intensity maximum
        - the exact praat command to use
    + centre of gravity
        - p=2, p=1 and/or p=⅔

There's also a newish option on the page for specifying a custom script, which could look something like this:
```
## This is a Praat Script:

# get centre of gravity and spread from spectrum
spectrum = To Spectrum... yes
# filter it
Filter (pass Hann band)... 1000 22000 100
# get centre of gravity
cog = Get centre of gravity... 2
# extract the result back out into a CSV column called 'cog'
print 'cog:0' 'newline$'
# tidy up objects
select spectrum
Remove
```

There are a few options for how to specify these from within R. The easiest would be to decide that only a limited set of options is initially available, something like this:

```
## load a results list
results <- read.csv("results.csv", header=T)

## get F1, F2, and mean pitch for all the segments
praat.results <- labbcat.processWithPraat(labbcat, 
  results$MatchId, results$segments.start, results$segments.end, window.offset=0.5, 
  formant.F1=TRUE, formant.F2=TRUE, 
  pitch.mean=TRUE)
```

Alternatively, or later on, more possible parameters could maybe be added:

```
## get F1 and F2 from 3 points, with max formant = 5000 for females and 5500 for males
praat.results <- labbcat.processWithPraat(labbcat, 
  results$MatchId, results$segments.start, results$segments.end, window.offset=0.5, 
  formant.F1=TRUE, formant.F2=TRUE, 
  formant.sample.points=c(0.25, 0.5, 0.75), 
  formant.max.female=5000, formant.max.male=5500)
```

I'm not sure whether it makes sense to include the possibility of a custom Praat script, but if so, it could be another possible parameter:

```
## Read the Praat script from a file
script.file <- "coolStuff.praat""
script <- readChar(script.file, file.info(script.file)$size)

## Run the Praat script on each token
praat.results <- labbcat.processWithPraat(labbcat, 
  results$MatchId, results$segments.start, results$segments.end, window.offset=0.5, 
  praat.script=script)
```

**Comment by robertfromont** [comment]
A better idea (thanks @djvill) might be to have the function only accept a script, and have another function, or series of functions, that do the work of created templated scripts:
```
## Create a script for extracting F1 and F2 from 3 points during the segment
script <- formantScript(F1=TRUE, F2=TRUE, sample.points=c(0.25, 0.5, 0.75))

## Run the Praat script on each token
praat.results <- processWithPraat(labbcat, 
  results$MatchId, results$segments.start, results$segments.end, window.offset=0.5, 
  praat.script=script)
```

**Comment by robertfromont** [comment]
Or (again from the brain of @djvill) implement a [Shiny Gadget](https://shiny.rstudio.com/articles/gadgets.html).

**Comment by robertfromont** [comment]
This has been implemented for release in 0.5-1 - e.g.
``` R
## get F1 and F2 for the mid point of the vowel
formants <- processWithPraat(
       labbcat.url,
       results$MatchId, results$Target.segments.start, results$Target.segments.end,
       praatScriptFormants(),
       no.progress=TRUE)

## get first 3 formants at three points during the sample, the mean, min, and max
## pitch, the max intensity, and the CoG using powers 1 and 2 
acoustic.measurements <- processWithPraat(
       labbcat.url,
        results$MatchId, results$Target.segments.start, results$Target.segments.end,
       paste(
           praatScriptFormants(c(1,2,3), c(0.25,0.5,0.75)),
           praatScriptPitch(get.mean=TRUE, get.minimum=TRUE, get.maximum=TRUE),
           praatScriptIntensity(),
           praatScriptCentreOfGravity(powers=c(1.0,2.0))),
       window.offset=0.5)
```
